test/performance/odp_l2fwd_perf.c

```diff
@@ -0,0 +1,1472 @@
+		}
+	}
+
+	return check_options(config);
+}
+
+static parse_result_t setup_program(int argc, char **argv, prog_config_t *config)
+{
+	struct sigaction action = { .sa_handler = terminate };
+
+	odp_atomic_init_u32(&config->is_running, 1U);
+
+	if (sigemptyset(&action.sa_mask) == -1 || sigaddset(&action.sa_mask, SIGINT) == -1 ||
+	    sigaddset(&action.sa_mask, SIGTERM) == -1 ||
+	    sigaddset(&action.sa_mask, SIGHUP) == -1 || sigaction(SIGINT, &action, NULL) == -1 ||
+	    sigaction(SIGTERM, &action, NULL) == -1 || sigaction(SIGHUP, &action, NULL) == -1) {
+		ODPH_ERR("Error installing signal handler\n");
+		return PRS_NOK;
+	}
+
+	return parse_options(argc, argv, config);
+}
+
+static void bind_destinations(pktio_t *pktios, uint8_t num)
+{
+	pktio_t *pktio_src, *pktio_dst;
+
+	for (uint8_t i = 0U; i < num; ++i) {
+		pktio_src = &pktios[i];
+
+		for (uint8_t j = 0U; j < num; ++j) {
+			pktio_dst = &pktios[j];
+
+			if (pktio_src->dst_name != NULL &&
+			    strcmp(pktio_src->dst_name, pktio_dst->name) == 0) {
+				pktio_src->dst = pktio_dst;
+				/* Copy the actual destination MAC as the 'dst_mac' that was
+				 * earlier saved to source's 'src_mac'. */
+				pktio_dst->dst_mac = pktio_src->src_mac;
+				break;
+			}
+		}
+	}
+}
+
+static odp_bool_t is_destination(const pktio_t *pktio, pktio_t **pktios, uint8_t num)
+{
+	for (uint8_t i = 0U; i < num; ++i)
+		if (pktio == pktios[i]->dst)
+			return true;
+
+	return false;
+}
+
+static void bind_workers(prog_config_t *config)
+{
+	const uint32_t num_workers = config->num_workers, num_ifs = config->num_ifs;
+	uint32_t max, min, i = 0U, j = 0U, *work_idx_ptr, *pktio_idx_ptr;
+	worker_config_t *worker;
+	pktio_t *pktio;
+
+	if (num_workers >= num_ifs) {
+		max = num_workers;
+		min = num_ifs;
+		work_idx_ptr = &i;
+		pktio_idx_ptr = &j;
+	} else {
+		max = num_ifs;
+		min = num_workers;
+		work_idx_ptr = &j;
+		pktio_idx_ptr = &i;
+	}
+
+	/* Assign workers to packet I/Os. Based on worker and packet I/O counts, the outer loop
+	 * will be looping the one with greater count and the inner with lesser count. */
+	for (; i < max; ++i) {
+		if (j == min)
+			j = 0U;
+
+		worker = &config->worker_config[*work_idx_ptr];
+		worker->pktios[worker->num_ifs++] = &config->pktios[*pktio_idx_ptr];
+		++j;
+	}
+
+	/* Check how many workers will end up polling a certain packet I/O input and increase input
+	 * queue count for that packet I/O accordingly. */
+	for (i = 0U; i < num_workers; ++i) {
+		worker = &config->worker_config[i];
+
+		for (j = 0U; j < worker->num_ifs; ++j)
+			++worker->pktios[j]->num_in_qs;
+	}
+
+	/* Check how many workers will end up outputting through a certain packet I/O and increase
+	 * output queue count for that packet I/O accordingly. */
+	for (i = 0U; i < num_ifs; ++i) {
+		pktio = &config->pktios[i];
+
+		for (j = 0U; j < num_workers; ++j) {
+			worker = &config->worker_config[j];
+
+			if (is_destination(pktio, worker->pktios, worker->num_ifs))
+				++pktio->num_out_qs;
+		}
+	}
```

Comment:
> So the most efficient option would be to create num_workers TX queues to each interface so ODP_PKTIO_OP_MT_UNSAFE can be used.

But if we want to preserve packet order (and why would we use ordered scheduling if we did not?), then we cannot send packets to worker specific TX queues but to flow specific TX queues and in that case multiple workers would end up sending to the same queue. I did not actually look at the code under review, but this comment sounded slightly dubious.

Reply:
This test application, like the original l2fwd app, doesn't care about packet order (direct output is always used). This could perhaps be mentioned in the application description / help print. Supporting ordered input queues can still be useful for measuring implementation overhead.

We have another test application `odp_pktio_ordered`, which maintains packet order.